packages/network-controller/src/NetworkController.ts

```diff
@@ -1374,6 +1374,10 @@ export class NetworkController extends BaseController<
    * @returns networkClientId of the network configuration with the given chainId
    */
   findNetworkClientIdByChainId(chainId: Hex): NetworkClientId {
+    if (this.state.providerConfig.chainId === chainId) {
+      return this.state.providerConfig.id ?? this.state.providerConfig.type;
```

Comment:
I initially tried `return this.state.selectedNetworkClientId`.  But I found that `providerConfig` and `selectedNetworkClientId` can be on different chains during a switch:

![image](https://github.com/MetaMask/core/assets/3500406/6da210bf-c93d-4966-bbe2-38421217d362)

Reply:
Makes sense. A `NetworkController:stateChange` event for the providerConfig changing fires before a stateChange event for the selectedNetworkClientId changing. 

I feel the order of events we fire is incorrect. Currently when the active network changes events are fired in the following order:
1. `NetworkController:stateChange` for the providerConfig change
2. `NetworkController:networkWillChange`
3. `NetworkController:stateChange` for the selectedNetworkClientId change and default networksMetadata
4. `NetworkController:networkDidChange`
5. `NetworkController:stateChange` for updating networksMetadata with EIP1559 compatibility
6. `NetworkController:infuraIsUnblocked`

It seems like steps 1 and 3 should be combined and fired after step 2. So it should look like this:
1. `NetworkController:networkWillChange`
2. `NetworkController:stateChange` for the providerConfig, selectedNetworkClientId, and default networksMetadata
3. `NetworkController:networkDidChange`
4. `NetworkController:stateChange` for updating networksMetadata with EIP1559 compatibility
5. `NetworkController:infuraIsUnblocked`


@Gudahtt @mcmire @adonesky1 @shanejonas @BelfordZ  thoughts?

Reply:
I don't think it matters when the `stateChange` occurs, because clients shouldn't be using that event to know whether the network changes, they should be using `networkDidChange` — and that already works today. But grouping like state changes together in the same update does make sense, so that we can guarantee that `providerConfig` matches `selectedNetworkClientId`.